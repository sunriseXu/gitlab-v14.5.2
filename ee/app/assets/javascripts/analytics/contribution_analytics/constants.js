import dateFormat from 'dateformat';
import { dateFormats } from '~/analytics/shared/constants';
import { getDateInPast, nMonthsBefore } from '~/lib/utils/datetime_utility';
import { s__ } from '~/locale';

export const CHART_HEIGHT = 350;
export const INNER_CHART_HEIGHT = 200;
export const CHART_X_AXIS_ROTATE = 45;
export const CHART_X_AXIS_NAME_TOP_PADDING = 55;

export const DATE_OPTIONS = [
  {
    text: s__('ContributionAnalytics|Last week'),
    value: dateFormat(getDateInPast(new Date(), 7), dateFormats.isoDate),
  },
  {
    text: s__('ContributionAnalytics|Last month'),
    value: dateFormat(nMonthsBefore(new Date(), 1), dateFormats.isoDate),
  },
  {
    text: s__('ContributionAnalytics|Last 3 months'),
    value: dateFormat(nMonthsBefore(new Date(), 3), dateFormats.isoDate),
  },
];
