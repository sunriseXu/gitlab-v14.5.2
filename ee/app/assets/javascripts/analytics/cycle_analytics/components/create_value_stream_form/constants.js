import { __, s__, sprintf } from '~/locale';

export const NAME_MAX_LENGTH = 100;

export const i18n = {
  FORM_TITLE: s__('CreateValueStreamForm|Create Value Stream'),
  EDIT_FORM_TITLE: s__('CreateValueStreamForm|Edit Value Stream'),
  EDIT_FORM_ACTION: s__('CreateValueStreamForm|Save value stream'),
  FORM_CREATED: s__("CreateValueStreamForm|'%{name}' Value Stream created"),
  FORM_EDITED: s__("CreateValueStreamForm|'%{name}' Value Stream saved"),
  RECOVER_HIDDEN_STAGE: s__('CreateValueStreamForm|Recover hidden stage'),
  RESTORE_HIDDEN_STAGE: s__('CreateValueStreamForm|Restore stage'),
  RESTORE_DEFAULTS: s__('CreateValueStreamForm|Restore defaults'),
  RECOVER_STAGE_TITLE: s__('CreateValueStreamForm|Default stages'),
  RECOVER_STAGES_VISIBLE: s__('CreateValueStreamForm|All default stages are currently visible'),
  SELECT_START_EVENT: s__('CreateValueStreamForm|Select start event'),
  SELECT_END_EVENT: s__('CreateValueStreamForm|Select end event'),
  FORM_FIELD_NAME_LABEL: s__('CreateValueStreamForm|Value Stream name'),
  FORM_FIELD_NAME_PLACEHOLDER: s__('CreateValueStreamForm|Enter value stream name'),
  FORM_FIELD_STAGE_NAME_PLACEHOLDER: s__('CreateValueStreamForm|Enter stage name'),
  FORM_FIELD_START_EVENT: s__('CreateValueStreamForm|Start event'),
  FORM_FIELD_START_EVENT_LABEL: s__('CreateValueStreamForm|Start event label'),
  FORM_FIELD_END_EVENT: s__('CreateValueStreamForm|End event'),
  FORM_FIELD_END_EVENT_LABEL: s__('CreateValueStreamForm|End event label'),
  DEFAULT_FIELD_START_EVENT_LABEL: s__('CreateValueStreamForm|Start event: '),
  DEFAULT_FIELD_END_EVENT_LABEL: s__('CreateValueStreamForm|End event: '),
  BTN_UPDATE_STAGE: s__('CreateValueStreamForm|Update stage'),
  BTN_ADD_STAGE: s__('CreateValueStreamForm|Add stage'),
  BTN_ADD_ANOTHER_STAGE: s__('CreateValueStreamForm|Add another stage'),
  TITLE_EDIT_STAGE: s__('CreateValueStreamForm|Editing stage'),
  TITLE_ADD_STAGE: s__('CreateValueStreamForm|New stage'),
  BTN_CANCEL: __('Cancel'),
  STAGE_INDEX: s__('CreateValueStreamForm|Stage %{index}'),
  HIDDEN_DEFAULT_STAGE: s__('CreateValueStreamForm|%{name} (default)'),
  TEMPLATE_DEFAULT: s__('CreateValueStreamForm|Create from default template'),
  TEMPLATE_BLANK: s__('CreateValueStreamForm|Create from no template'),
  ISSUE_STAGE_END: s__('CreateValueStreamForm|Issue stage end'),
  PLAN_STAGE_START: s__('CreateValueStreamForm|Plan stage start'),
  CODE_STAGE_START: s__('CreateValueStreamForm|Code stage start'),
  DEFAULT_STAGE_FEATURES: s__(
    'CreateValueStreamForm|Default stages can only be hidden or re-ordered',
  ),
};

export const ERRORS = {
  VALUE_STREAM_NAME_MIN_LENGTH: s__('CreateValueStreamForm|Name is required'),
  STAGE_NAME_MIN_LENGTH: s__('CreateValueStreamForm|Stage name is required'),
  MAX_LENGTH: sprintf(s__('CreateValueStreamForm|Maximum length %{maxLength} characters'), {
    maxLength: NAME_MAX_LENGTH,
  }),
  START_EVENT_REQUIRED: s__('CreateValueStreamForm|Please select a start event first'),
  END_EVENT_REQUIRED: s__('CreateValueStreamForm|Please select an end event'),
  STAGE_NAME_EXISTS: s__('CreateValueStreamForm|Stage name already exists'),
  INVALID_EVENT_PAIRS: s__(
    'CreateValueStreamForm|Start event changed, please select a valid end event',
  ),
};

export const STAGE_SORT_DIRECTION = {
  UP: 'UP',
  DOWN: 'DOWN',
};

export const formFieldKeys = [
  'id',
  'name',
  'startEventIdentifier',
  'endEventIdentifier',
  'startEventLabelId',
  'endEventLabelId',
];

export const editableFormFieldKeys = [
  ...formFieldKeys,
  'hidden',
  'description',
  'title',
  'legend',
  'custom',
];

export const defaultFields = formFieldKeys.reduce((acc, field) => ({ ...acc, [field]: null }), {});
export const defaultErrors = formFieldKeys.reduce((acc, field) => ({ ...acc, [field]: [] }), {});

export const defaultCustomStageFields = { ...defaultFields, custom: true };

export const PRESET_OPTIONS_DEFAULT = 'default';
export const PRESET_OPTIONS_BLANK = 'blank';
export const PRESET_OPTIONS = [
  {
    text: i18n.TEMPLATE_DEFAULT,
    value: PRESET_OPTIONS_DEFAULT,
  },
  {
    text: i18n.TEMPLATE_BLANK,
    value: PRESET_OPTIONS_BLANK,
  },
];

// These events can only be set on the back end, they are used in the
// initial configuration of some default stages, but should not be
// selectable by users via the form, they are added here only for display
// purposes when we are editing a default value stream
export const ADDITIONAL_DEFAULT_STAGE_EVENTS = [
  {
    identifier: 'issue_stage_end',
    name: i18n.ISSUE_STAGE_END,
  },
  {
    identifier: 'plan_stage_start',
    name: i18n.PLAN_STAGE_START,
  },
  {
    identifier: 'code_stage_start',
    name: i18n.CODE_STAGE_START,
  },
];
