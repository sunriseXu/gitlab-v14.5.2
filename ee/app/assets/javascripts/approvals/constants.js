import { __, s__ } from '~/locale';

export const TYPE_USER = 'user';
export const TYPE_GROUP = 'group';
export const TYPE_HIDDEN_GROUPS = 'hidden_groups';

export const BRANCH_FETCH_DELAY = 250;
export const ALL_BRANCHES = {
  id: null,
  name: __('All branches'),
};

export const RULE_TYPE_FALLBACK = 'fallback';
export const RULE_TYPE_REGULAR = 'regular';
export const RULE_TYPE_REPORT_APPROVER = 'report_approver';
export const RULE_TYPE_CODE_OWNER = 'code_owner';
export const RULE_TYPE_ANY_APPROVER = 'any_approver';
export const RULE_NAME_ANY_APPROVER = 'All Members';

export const VULNERABILITY_CHECK_NAME = 'Vulnerability-Check';
export const LICENSE_CHECK_NAME = 'License-Check';
export const COVERAGE_CHECK_NAME = 'Coverage-Check';

export const REPORT_TYPE_LICENSE_SCANNING = 'license_scanning';
export const REPORT_TYPE_VULNERABILITY = 'vulnerability';
export const REPORT_TYPE_CODE_COVERAGE = 'code_coverage';

export const APPROVAL_RULE_CONFIGS = {
  [VULNERABILITY_CHECK_NAME]: {
    title: s__('SecurityApprovals|Vulnerability-Check'),
    popoverText: s__(
      'SecurityApprovals|A merge request approval is required when a security report contains a new vulnerability.',
    ),
    documentationText: s__('SecurityApprovals|Learn more about Vulnerability-Check'),
    reportType: REPORT_TYPE_VULNERABILITY,
  },
  [LICENSE_CHECK_NAME]: {
    title: s__('SecurityApprovals|License-Check'),
    popoverText: s__(
      'SecurityApprovals|A merge request approval is required when the license compliance report contains a denied license.',
    ),
    documentationText: s__('SecurityApprovals|Learn more about License-Check'),
    reportType: REPORT_TYPE_LICENSE_SCANNING,
  },
  [COVERAGE_CHECK_NAME]: {
    title: s__('SecurityApprovals|Coverage-Check'),
    popoverText: s__(
      'SecurityApprovals|A merge request approval is required when test coverage declines.',
    ),
    documentationText: s__('SecurityApprovals|Learn more about Coverage-Check'),
    reportType: REPORT_TYPE_CODE_COVERAGE,
  },
};

export const APPROVALS_HELP_PATH = 'user/project/merge_requests/approvals/settings';

export const APPROVAL_SETTINGS_I18N = {
  learnMore: __('Learn more.'),
  approvalSettingsHeader: __('Approval settings'),
  approvalSettingsDescription: __('Define how approval rules are applied to merge requests.'),
  saveChanges: __('Save changes'),
  loadingErrorMessage: s__(
    'ApprovalSettings|There was an error loading merge request approval settings.',
  ),
  savingErrorMessage: s__(
    'ApprovalSettings|There was an error updating merge request approval settings.',
  ),
  savingSuccessMessage: s__('ApprovalSettings|Merge request approval settings have been updated.'),
  lockedByAdmin: s__(
    'ApprovalSettings|This setting is configured at the instance level and can only be changed by an administrator.',
  ),
  lockedByGroupOwner: s__(
    'ApprovalSettings|This setting is configured in %{groupName} and can only be changed by an administrator or group owner.',
  ),
};

export const PROJECT_APPROVAL_SETTINGS_LABELS_I18N = {
  authorApprovalLabel: s__('ApprovalSettings|Prevent approval by author.'),
  preventMrApprovalRuleEditLabel: s__(
    'ApprovalSettings|Prevent editing approval rules in merge requests.',
  ),
  preventCommittersApprovalLabel: s__(
    'ApprovalSettings|Prevent approvals by users who add commits.',
  ),
  requireUserPasswordLabel: s__('ApprovalSettings|Require user password to approve.'),
  removeApprovalsOnPushLabel: s__(
    'ApprovalSettings|Remove all approvals when commits are added to the source branch.',
  ),
};

export const GROUP_APPROVAL_SETTINGS_LABELS_I18N = {
  ...PROJECT_APPROVAL_SETTINGS_LABELS_I18N,
  preventMrApprovalRuleEditLabel: s__(
    'ApprovalSettings|Prevent editing approval rules in projects and merge requests.',
  ),
};

export const APPROVAL_DIALOG_I18N = {
  form: {
    approvalsRequiredLabel: s__('ApprovalRule|Approvals required'),
    approvalTypeLabel: s__('ApprovalRule|Approver Type'),
    approversLabel: s__('ApprovalRule|Add approvers'),
    nameLabel: s__('ApprovalRule|Rule name'),
    nameDescription: s__('ApprovalRule|Examples: QA, Security.'),
    protectedBranchLabel: s__('ApprovalRule|Target branch'),
    protectedBranchDescription: __(
      'Apply this approval rule to all branches or a specific protected branch.',
    ),
    scannersLabel: s__('ApprovalRule|Security scanners'),
    scannersSelectLabel: s__('ApprovalRule|Select scanners'),
    scannersDescription: s__(
      'ApprovalRule|Apply this approval rule to consider only the selected security scanners.',
    ),
    selectAllLabel: s__('ApprovalRule|Select All'),
    allScannersSelectedLabel: s__('ApprovalRule|All scanners'),
    multipleSelectedLabel: s__('ApprovalRule|%{firstLabel} +%{numberOfAdditionalLabels} more'),
    vulnerabilitiesAllowedLabel: s__('ApprovalRule|Vulnerabilities allowed'),
    vulnerabilitiesAllowedDescription: s__(
      'ApprovalRule|Number of vulnerabilities allowed before approval rule is triggered.',
    ),
    vulnerabilityStatesLabel: s__('ApprovalRule|Vulnerability states'),
    vulnerabilityStatesDescription: s__(
      'ApprovalRule|Apply this approval rule to consider only the selected vulnerability states.',
    ),
    vulnerabilityStatesSelectLabel: s__('ApprovalRule|Select vulnerability states'),
    allVulnerabilityStatesSelectedLabel: s__('ApprovalRule|All vulnerability states'),
    severityLevelsLabel: s__('ApprovalRule|Severity levels'),
    severityLevelsDescription: s__(
      'ApprovalRule|Apply this approval rule to consider only the selected severity levels.',
    ),
    severityLevelsSelectLabel: s__('ApprovalRule|Select severity levels'),
    allSeverityLevelsSelectedLabel: s__('ApprovalRule|All severity levels'),
  },
  validations: {
    approvalsRequiredNegativeNumber: __('Please enter a non-negative number'),
    approvalsRequiredNotNumber: __('Please enter a valid number'),
    approvalsRequiredMinimum: __(
      'Please enter a number greater than %{number} (from the project settings)',
    ),
    approversRequired: __('Please select and add a member'),
    branchesRequired: __('Please select a valid target branch'),
    ruleNameTaken: __('Rule name is already taken.'),
    ruleNameMissing: __('Please provide a name'),
    scannersRequired: s__('ApprovalRule|Please select at least one security scanner'),
    vulnerabilitiesAllowedMinimum: s__(
      'ApprovalRule|Please enter a number equal or greater than zero',
    ),
    severityLevelsRequired: s__('ApprovalRule|Please select at least one severity level'),
    vulnerabilityStatesRequired: s__('ApprovalRule|Please select at least one vulnerability state'),
  },
};

export const APPROVAL_VULNERABILITY_STATES = {
  newly_detected: s__('ApprovalRule|Newly detected'),
  detected: s__('ApprovalRule|Previously detected'),
  confirmed: s__('ApprovalRule|Confirmed'),
  dismissed: s__('ApprovalRule|Dismissed'),
  resolved: s__('ApprovalRule|Resolved'),
};
