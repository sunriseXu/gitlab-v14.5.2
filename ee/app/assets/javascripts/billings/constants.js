export const TABLE_TYPE_DEFAULT = 'default';
export const TABLE_TYPE_FREE = 'free';
export const TABLE_TYPE_TRIAL = 'trial';

export const DAYS_FOR_RENEWAL = 15;
