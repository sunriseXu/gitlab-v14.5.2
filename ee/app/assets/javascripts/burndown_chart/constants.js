export { Namespace } from '../iterations/constants';

export const Unit = {
  count: 'count',
  weight: 'weight',
};
