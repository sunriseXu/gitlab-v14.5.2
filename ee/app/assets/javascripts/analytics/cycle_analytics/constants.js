import { getGroupValueStreamMetrics } from 'ee/api/analytics_api';
import { METRIC_TYPE_SUMMARY, METRIC_TYPE_TIME_SUMMARY } from '~/api/analytics_api';
import { OVERVIEW_STAGE_ID } from '~/cycle_analytics/constants';
import { __ } from '~/locale';

export const EVENTS_LIST_ITEM_LIMIT = 50;

export const TASKS_BY_TYPE_SUBJECT_ISSUE = 'Issue';
export const TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST = 'MergeRequest';
export const TASKS_BY_TYPE_MAX_LABELS = 15;

export const TASKS_BY_TYPE_SUBJECT_FILTER_OPTIONS = {
  [TASKS_BY_TYPE_SUBJECT_ISSUE]: __('Issues'),
  [TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST]: __('Merge Requests'),
};

export const TASKS_BY_TYPE_FILTERS = {
  SUBJECT: 'SUBJECT',
  LABEL: 'LABEL',
};

export const DEFAULT_VALUE_STREAM_ID = 'default';

export const FETCH_VALUE_STREAM_DATA = 'fetchValueStreamData';

export const OVERVIEW_STAGE_CONFIG = {
  id: OVERVIEW_STAGE_ID,
  title: __('Overview'),
  icon: 'home',
};

export const METRICS_REQUESTS = [
  {
    endpoint: METRIC_TYPE_TIME_SUMMARY,
    request: getGroupValueStreamMetrics,
    name: __('time summary'),
  },
  {
    endpoint: METRIC_TYPE_SUMMARY,
    request: getGroupValueStreamMetrics,
    name: __('recent activity'),
  },
];
