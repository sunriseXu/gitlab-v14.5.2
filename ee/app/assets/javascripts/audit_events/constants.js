import { s__ } from '~/locale';
import { OPERATOR_IS_ONLY } from '~/vue_shared/components/filtered_search_bar/constants';
import GroupToken from './components/tokens/group_token.vue';
import MemberToken from './components/tokens/member_token.vue';
import ProjectToken from './components/tokens/project_token.vue';
import UserToken from './components/tokens/user_token.vue';

const DEFAULT_TOKEN_OPTIONS = {
  operators: OPERATOR_IS_ONLY,
  unique: true,
};

// Due to the i18n eslint rule we can't have a capitalized string even if it is a case-aware URL param
/* eslint-disable @gitlab/require-i18n-strings */
export const ENTITY_TYPES = {
  USER: 'User',
  AUTHOR: 'Author',
  GROUP: 'Group',
  PROJECT: 'Project',
};
/* eslint-enable @gitlab/require-i18n-strings */

export const AUDIT_FILTER_CONFIGS = [
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'user',
    title: s__('AuditLogs|User Events'),
    type: 'user',
    entityType: ENTITY_TYPES.USER,
    token: UserToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'user',
    title: s__('AuditLogs|Member Events'),
    type: 'member',
    entityType: ENTITY_TYPES.AUTHOR,
    token: MemberToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'bookmark',
    title: s__('AuditLogs|Project Events'),
    type: 'project',
    entityType: ENTITY_TYPES.PROJECT,
    token: ProjectToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'group',
    title: s__('AuditLogs|Group Events'),
    type: 'group',
    entityType: ENTITY_TYPES.GROUP,
    token: GroupToken,
  },
];

export const AVAILABLE_TOKEN_TYPES = AUDIT_FILTER_CONFIGS.map((token) => token.type);

export const MAX_DATE_RANGE = 31;

export const SAME_DAY_OFFSET = 1;

// This creates a date with zero time, making it simpler to match to the query date values
export const CURRENT_DATE = new Date(new Date().toDateString());
