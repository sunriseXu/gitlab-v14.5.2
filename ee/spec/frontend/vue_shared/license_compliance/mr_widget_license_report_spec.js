import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import LicenseManagement from 'ee/vue_shared/license_compliance/mr_widget_license_report.vue';
import { stubComponent } from 'helpers/stub_component';
import { TEST_HOST } from 'spec/test_constants';
import ReportItem from '~/reports/components/report_item.vue';
import ReportSection from '~/reports/components/report_section.vue';
import { LOADING, ERROR, SUCCESS } from '~/reports/constants';
import {
  approvedLicense,
  blacklistedLicense,
  licenseReport as licenseReportMock,
  generateReportGroup,
} from './mock_data';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('License Report MR Widget', () => {
  const apiUrl = `${TEST_HOST}/license_management`;
  const licenseComplianceDocsPath = `${TEST_HOST}/path/to/security/approvals/help`;
  let wrapper;

  const defaultState = {
    managedLicenses: [approvedLicense, blacklistedLicense],
    currentLicenseInModal: licenseReportMock[0],
    isLoadingManagedLicenses: true,
  };

  const defaultGetters = {
    isLoading() {
      return false;
    },
    licenseReport() {
      return licenseReportMock;
    },
    licenseSummaryText() {
      return 'FOO';
    },
    reportContainsBlacklistedLicense() {
      return false;
    },
    licenseReportGroups() {
      return [];
    },
  };

  const defaultProps = {
    loadingText: 'LOADING',
    errorText: 'ERROR',
    licensesApiPath: `${TEST_HOST}/parsed_license_report.json`,
    approvalsApiPath: `${TEST_HOST}/path/to/approvals`,
    canManageLicenses: true,
    licenseManagementSettingsPath: `${TEST_HOST}/lm_settings`,
    fullReportPath: `${TEST_HOST}/path/to/the/full/report`,
    apiUrl,
    licenseComplianceDocsPath,
  };

  const defaultActions = {
    setAPISettings: () => {},
    fetchManagedLicenses: () => {},
    fetchParsedLicenseReport: () => {},
    fetchLicenseCheckApprovalRule: () => {},
  };

  const mountComponent = ({
    props = defaultProps,
    getters = defaultGetters,
    state = defaultState,
    actions = defaultActions,
    stubs = { ReportSection },
  } = {}) => {
    const store = new Vuex.Store({
      modules: {
        licenseManagement: {
          namespaced: true,
          state,
          getters,
          actions,
        },
      },
    });
    wrapper = shallowMount(LicenseManagement, {
      localVue,
      propsData: props,
      store,
      stubs,
    });
  };

  const findAllReportItems = () => wrapper.findAll(ReportItem);

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('computed', () => {
    describe('hasLicenseReportIssues', () => {
      it('should be false, if the report is empty', () => {
        const getters = {
          ...defaultGetters,
          licenseReport() {
            return [];
          },
        };
        mountComponent({ getters });

        expect(wrapper.vm.hasLicenseReportIssues).toBe(false);
      });

      it('should be true, if the report is not empty', () => {
        mountComponent();

        expect(wrapper.vm.hasLicenseReportIssues).toBe(true);
      });
    });

    describe('licenseReportStatus', () => {
      it('should be `LOADING`, if the report is loading', () => {
        const getters = {
          ...defaultGetters,
          isLoading() {
            return true;
          },
        };
        mountComponent({ getters });

        expect(wrapper.vm.licenseReportStatus).toBe(LOADING);
      });

      it('should be `ERROR`, if the report is has an error', () => {
        const state = { ...defaultState, loadLicenseReportError: new Error('test') };
        mountComponent({ state });

        expect(wrapper.vm.licenseReportStatus).toBe(ERROR);
      });

      it('should be `SUCCESS`, if the report is successful', () => {
        mountComponent();

        expect(wrapper.vm.licenseReportStatus).toBe(SUCCESS);
      });
    });

    describe('showActionButtons', () => {
      const { fullReportPath, licenseManagementSettingsPath, ...otherProps } = defaultProps;

      it('should be true if fullReportPath AND licenseManagementSettingsPath prop are provided', () => {
        const props = { ...otherProps, fullReportPath, licenseManagementSettingsPath };
        mountComponent({ props });

        expect(wrapper.vm.showActionButtons).toBe(true);
      });

      it('should be true if only licenseManagementSettingsPath is provided', () => {
        const props = { ...otherProps, fullReportPath: null, licenseManagementSettingsPath };
        mountComponent({ props });

        expect(wrapper.vm.showActionButtons).toBe(true);
      });

      it('should be true if only fullReportPath is provided', () => {
        const props = {
          ...otherProps,
          fullReportPath,
          licenseManagementSettingsPath: null,
        };
        mountComponent({ props });

        expect(wrapper.vm.showActionButtons).toBe(true);
      });

      it('should be false if fullReportPath and licenseManagementSettingsPath prop are not provided', () => {
        const props = {
          ...otherProps,
          fullReportPath: null,
          licenseManagementSettingsPath: null,
        };
        mountComponent({ props });

        expect(wrapper.vm.showActionButtons).toBe(false);
      });
    });
  });

  describe('report section', () => {
    describe('report body', () => {
      it('should render correctly', () => {
        const mockReportGroups = [generateReportGroup()];

        mountComponent({
          getters: {
            ...defaultGetters,
            licenseReportGroups() {
              return mockReportGroups;
            },
          },
        });

        expect(wrapper.find({ ref: 'reportSectionBody' }).element).toMatchSnapshot();
      });

      it.each`
        givenStatuses                       | expectedNumberOfReportHeadings
        ${[]}                               | ${0}
        ${['failed', 'neutral']}            | ${2}
        ${['failed', 'neutral', 'success']} | ${3}
      `(
        'given reports for: $givenStatuses it has $expectedNumberOfReportHeadings report headings',
        ({ givenStatuses, expectedNumberOfReportHeadings }) => {
          const mockReportGroups = givenStatuses.map((status) => generateReportGroup({ status }));

          mountComponent({
            getters: {
              ...defaultGetters,
              licenseReportGroups() {
                return mockReportGroups;
              },
            },
          });

          expect(wrapper.findAll({ ref: 'reportHeading' })).toHaveLength(
            expectedNumberOfReportHeadings,
          );
        },
      );

      it.each([0, 1, 2])(
        'should include %d report items when section has that many licenses',
        (numberOfLicenses) => {
          const mockReportGroups = [
            generateReportGroup({
              numberOfLicenses,
            }),
          ];

          mountComponent({
            getters: {
              ...defaultGetters,
              licenseReportGroups() {
                return mockReportGroups;
              },
            },
          });

          expect(findAllReportItems()).toHaveLength(numberOfLicenses);
        },
      );

      it('renders the report items in the correct order', () => {
        const mockReportGroups = [
          generateReportGroup({ status: 'failed', numberOfLicenses: 1 }),
          generateReportGroup({ status: 'neutral', numberOfLicenses: 1 }),
          generateReportGroup({ status: 'success', numberOfLicenses: 1 }),
        ];

        mountComponent({
          getters: {
            ...defaultGetters,
            licenseReportGroups() {
              return mockReportGroups;
            },
          },
        });

        const allReportItems = findAllReportItems();
        mockReportGroups.forEach((group, index) => {
          expect(allReportItems.at(index).props('status')).toBe(group.status);
        });
      });
    });
  });

  describe('`View full report` button', () => {
    const selector = '[data-testid="full-report-button"]';

    it('should be rendered when fullReportPath prop is provided', () => {
      mountComponent();

      const linkEl = wrapper.find(selector);

      expect(linkEl.exists()).toBe(true);
      expect(linkEl.attributes('href')).toEqual(defaultProps.fullReportPath);
      expect(linkEl.text()).toBe('View full report');
    });

    it('should not be rendered when fullReportPath prop is not provided', () => {
      const props = { ...defaultProps, fullReportPath: null };
      mountComponent({ props });

      expect(wrapper.find(selector).exists()).toBe(false);
    });
  });

  describe('`Manage licenses` button', () => {
    const selector = '[data-testid="manage-licenses-button"]';

    it('should be rendered when licenseManagementSettingsPath prop is provided', () => {
      mountComponent();

      const linkEl = wrapper.find(selector);

      expect(linkEl.exists()).toBe(true);
      expect(linkEl.attributes('href')).toEqual(defaultProps.licenseManagementSettingsPath);
      expect(linkEl.text()).toBe('Manage licenses');
    });

    it('should not be rendered when licenseManagementSettingsPath prop is not provided', () => {
      const props = { ...defaultProps, licenseManagementSettingsPath: null };
      mountComponent({ props });

      expect(wrapper.find(selector).exists()).toBe(false);
    });

    it('has gl-mr-3 class when isCollapsbile is true', () => {
      mountComponent({
        stubs: {
          ReportSection: stubComponent(ReportSection, {
            template: `
              <div>
                <slot name="action-buttons" :is-collapsible="true" />
              </div>
            `,
          }),
        },
      });
      expect(wrapper.find(selector).classes()).toContain('gl-mr-3');
    });

    it('does not have gl-mr-3 class when isCollapsbile is false', () => {
      mountComponent({
        stubs: {
          ReportSection: stubComponent(ReportSection, {
            template: `
              <div>
                <slot name="action-buttons" :is-collapsible="false" />
              </div>
            `,
          }),
        },
      });
      expect(wrapper.find(selector).classes()).not.toContain('gl-mr-3');
    });
  });

  it('should init store after mount', () => {
    const actions = {
      setAPISettings: jest.fn(),
      fetchParsedLicenseReport: jest.fn(),
      fetchLicenseCheckApprovalRule: jest.fn(),
    };
    mountComponent({ actions });

    expect(actions.setAPISettings).toHaveBeenCalledWith(expect.any(Object), {
      apiUrlManageLicenses: apiUrl,
      licensesApiPath: defaultProps.licensesApiPath,
      approvalsApiPath: defaultProps.approvalsApiPath,
      canManageLicenses: true,
    });

    expect(actions.fetchParsedLicenseReport).toHaveBeenCalledWith(expect.any(Object), undefined);

    expect(actions.fetchLicenseCheckApprovalRule).toHaveBeenCalledWith(
      expect.any(Object),
      undefined,
    );
  });

  describe('approval status', () => {
    const findLicenseComplianceHelpLink = () =>
      wrapper.find('[data-testid="security-approval-help-link"]');

    it('does not show a link to security approval help page if report does not contain blacklisted licenses', () => {
      mountComponent();

      expect(findLicenseComplianceHelpLink().exists()).toBe(false);
    });

    it('shows a link to security approval help page if report contains blacklisted licenses', () => {
      const getters = {
        ...defaultGetters,
        reportContainsBlacklistedLicense() {
          return true;
        },
      };
      mountComponent({
        getters,
      });

      const licenseComplianceHelpLink = findLicenseComplianceHelpLink();

      expect(findLicenseComplianceHelpLink().exists()).toBe(true);
      expect(licenseComplianceHelpLink.attributes('href')).toBe(licenseComplianceDocsPath);
    });
  });
});
