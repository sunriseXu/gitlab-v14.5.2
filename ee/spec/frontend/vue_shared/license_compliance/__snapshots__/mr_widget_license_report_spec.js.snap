// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`License Report MR Widget report section report body should render correctly 1`] = `
<smart-virtual-list-stub
  class="report-block-container"
  length="2"
  remain="20"
  rtag="div"
  size="26"
  wclass="report-block-list my-1"
  wtag="ul"
>
  <li
    class="mx-1 mb-1"
  >
    <h2
      class="h5 m-0"
    >
      some-status group-name
    </h2>
     
    <p
      class="m-0"
    >
      some-status group-description
    </p>
  </li>
   
</smart-virtual-list-stub>
`;
